src/uptane.rs

```diff
@@ -17,47 +17,92 @@ pub struct Version {
     timestamp: u64
 }
 
-impl Version {
-    fn new() -> Self {
+impl Default for Version {
+    fn default() -> Self {
         Version { root: 0, targets: 0, snapshot: 0, timestamp: 0 }
     }
 }
 
 
 /// Software over the air updates using Uptane endpoints.
 pub struct Uptane {
-    uptane_cfg:  UptaneConfig,
-    gateway:     Url,
-    device_uuid: String,
-    version:     Version,
-    verifier:    Verifier,
+    uptane_cfg: UptaneConfig,
+    deviceid:   String,
+    version:    Version,
+    verifier:   Verifier,
+    serial:     String,
+    privkey:    PrivateKey,
+
+    pub fetch_root:    bool,
+    pub send_manifest: bool,
 }
 
-
 impl Uptane {
-    pub fn new(cfg: UptaneConfig, gateway: Url, device_uuid: String) -> Self {
-        Uptane {
-            uptane_cfg:  cfg,
-            gateway:     gateway,
-            device_uuid: device_uuid,
-            version:     Version::new(),
-            verifier:    Verifier::new(),
+    pub fn new(config: &Config) -> Result<Self, Error> {
+        let mut verifier = Verifier::new();
+        let der_key = read_file(&config.uptane.private_key_path)
+            .map_err(|err| Error::Client(format!("couldn't read uptane.private_key_path: {}", err)))?;
+
+        let root_path  = format!("{}/director/root.json", &config.uptane.metadata_path);
+        let fetch_root = if Path::new(&root_path).exists() {
+            let buf  = read_file(&root_path)?;
+            let meta = json::from_slice::<TufSigned>(&buf)?;
+            let root = json::from_value::<Root>(meta.signed.clone())?;
+            Self::add_root_keys(&mut verifier, &root);
+            false
+        } else {
+            true
+        };
+
+        Ok(Uptane {
+            uptane_cfg: config.uptane.clone(),
+            deviceid:   config.device.uuid.clone(),
+            version:    Version::default(),
+            verifier:   verifier,
+            serial:     config.uptane.primary_ecu_serial.clone(),
+            privkey:  PrivateKey {
+                // FIXME: keyid
+                keyid:   "e453c713367595e1a9e5c1de8b2c039fe4178094bdaf2d52b1993fdd1a76ee26".into(),
+                der_key: der_key
+            },
+            fetch_root:    fetch_root,
+            send_manifest: true,
+        })
+    }
+
+    pub fn initialize(&mut self, client: &Client) -> Result<(), Error> {
+        if self.send_manifest {
+            self.put_manifest(client)?;
+            self.send_manifest = false;
         }
+
+        if self.fetch_root {
+            let buf  = self.get_endpoint(client, true, "root.json")?;
+            let meta = json::from_slice::<TufSigned>(&buf)?;
+            let root = json::from_value::<Root>(meta.signed.clone())?;
+            Self::add_root_keys(&mut self.verifier, &root);
```

Comment:
Short answer: no. If we have downloaded a `root.json` then we assume it is valid. If there is a way to verify this then it would be good to add.